### DISCUSSION TITLE
Adstock & transformations

### DISCUSSION THREAD

**Discussion by ASevastkms** [facebookexperimental/Robyn#620] [Q&A] [answered]
Hello! As I understand it, Robin first transforms media variables by adding ADstock, and then saturation. Why this particular order? The only source I found is an article by Jin, Y., Wang, Y., Sun, Y., Chan, D. & Koehler, J. (2017). Bayesian methods for media mix modeling with carryover and shape effects, which actually provides a very concise rationale for choosing a sequence of transformations. Has anyone faced this issue? After all, these transformations are not commutative and the order of their application will change the result.

**Comment by gufengzhou** [comment, marked as answer]
This is a good question. We do have a reason behind the order of first adstock second saturation. For an always-on channel as an example, if you shut it down for a day, it probably will still convert on that day due to adstock. This means adstock is part of the saturation curve. Or to be precise, adstock is on the bottom of saturation curve, while actual spend sits on the higher part of the curve on top of adstock. This is exactly how robyn calculates the immediate & carryover effect for media channels.

**Reply by ASevastkms** [reply]
Thanks, but I don't quite understand your logic. If transforms were applied first and then adstock, then this "shutdown day" would also have a non-zero contribution from the media. So what's the difference?

**Reply by gufengzhou** [reply]
No it's adstock first, saturation second. They're both data transformation. 
Maybe this: we believe adstock can also saturate, same behaviour as raw spend. That's why we do adstock first, so that it gets the diminishing behaviour afterwards.

**Reply by shantisb** [reply]
Hey @gufengzhou super interesting. Quick question about transformations. When decomposing the baseline revenue Robyn firsts tries to explain what it can of revenue through seasonality, trend, holiday, etc., then media  variables are introduced next in the modeling process? Could that be an explanation as to why seasonality gets a higher attribution vs media (as opposed to other MMM packages like Meridian for example)?  So any revenue explained by these components is removed before assigning impact to media correct? 


Also since Robyn uses Michaelis Menten does that mean that from the revenue that was left (after factoring in non media variables) it imposes a natural cap and further limits how much of the remaining revenue can be assigned to media lowering media attribution over time? 

(well it wasn't that quick but really appreciate it @gufengzhou !)

**Reply by gufengzhou** [reply]
Thanks! Robyn decomposes trend/season first just to obtain the timeseries for these variables. Then trend/season/media etc will all go into model fitting to predict total sales. It doesn't favour trend/season this way. 
Robyn uses Hill function for saturation transformation, the other packages are using the same AFAIK. With all sigmoid-shape transformations, we'll get the diminishing behaviour that is the purpose of modeling saturation.

**Reply by shantisb** [reply]
@gufengzhou  but doesn’t Prophet provide a partial decomposition of the trend, seasonality, and holiday components? Robyn then re-fits those baseline components along with the media channels in the final MMM step to determine the final revenue attribution. While that initial decomposition doesn’t lock in the final magnitude of each baseline component, Prophet does give you a preliminary estimate—right? Couldn’t this be a potential source of over-attribution to the Prophet components?

Especially considering that Prophet incorporates factor (binary) variables into its objective function, and it uses ridge regularization—which penalizes large coefficients. So, when there’s a binary promo variable set to 1 for a specific week, Prophet might detect a revenue spike, but instead of assigning the impact strictly to that week, the regularization seems to be causing it to spread the effect across multiple weeks (even when the variable is off). I’ve been noticing revenue being attributed to weeks where the promo variable was actually set to 0.

Any help or clarification would be really appreciated! I’ve become super intrigued by MMM and just want to better understand what’s going on under the hood. I’ve been going through the code and trying to get a clearer picture of how everything works (hence the many questions haha).

Thank you so much!
